### PR TITLE
Provide Sharding Sphere as backend server.

### PR DESCRIPTION
**Pull request by tristaZero**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [ ] Bug fix
- [ ] New feature provided
- [ ] Improve performance

- Related issues
#4131
___

### New feature or improvement
- There are sharding algorithms for sharding Databases and Tables.
- `PreciseDatabaseShardingAlgorithm` is to shard DB with SQL `WHERE XXX = XX` or `WHERE XXX IN ()`
- `PreciseTableShardingAlgorithm` is to shard Tables with SQL `WHERE XXX = XX` or `WHERE XXX IN ()`
- `RangeDatabaseShardingAlgorithm` is to shard DB with SQL `XXX BETWEEN IN`
- `RangeTableShardingAlgorithm` is to shard Tables with SQL `XXX BETWEEN IN`

### CODE REVIEW

**Comment by tristaZero** [comment]
Hi @wu-sheng  
> Comments are required.

I am not sure which comment you mentioned, but i modifed my first comment. Do you think it makes sense?

> What are the steps of the further PRs? 

As [Schedule](https://github.com/apache/skywalking/issues/4131#issuecomment-587302393) said, the next jobs are `Provide configuration of Sharding Proxy for Skywalking.` and `Provide e2e tests of Skywalking.`

So maybe it is the time we'd better ask @dmsolr 's for help.

**Comment by wu-sheng** [comment]
I mean comments of codes. You missed that. Are you planning to wait for @dmsolr to provide e2e? Even yes, should you or he provide the sharding configuration?

**Comment by dmsolr** [comment]
Hi, we need to setup e2e for shardingsphere before the main features to merge, right?

**Comment by tristaZero** [comment]
> Hi, we need to setup e2e for shardingsphere before the main features to merge, right?

Yes, after this PR is merged, you can move farword. And [here](https://shardingsphere.apache.org/document/current/cn/quick-start/sharding-proxy-quick-start/) is introduction of ShardingProxy.

**Comment by tristaZero** [comment]
> I mean comments of codes. You missed that. Are you planning to wait for @dmsolr to provide e2e? Even yes, should you or he provide the sharding configuration?

Comments have been added.

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Collection;
+
+/**
+ * Precise sharding algorithm to shard database.
+ * 
+ * Sharding Sphere will use it to handle SQL with equal or in query condition.
+ */
+public final class PreciseDatabaseShardingAlgorithm implements PreciseShardingAlgorithm<String> {
+    
+    private final static int DB_NUMBER = 4;
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java]: Is this static? How should user inject the value?

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+
+package org.apache.skywalking.oap.server.storage.plugin.shardingsphere;
+
+import org.apache.shardingsphere.api.sharding.standard.PreciseShardingAlgorithm;
+import org.apache.shardingsphere.api.sharding.standard.PreciseShardingValue;
+
+import java.util.Collection;
+
+/**
+ * Precise sharding algorithm to shard database.
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java]: What does this sentence mean? `Precise` is an adj.. Could you re-organize?

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseTableShardingAlgorithm.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import java.util.Collection;
+
+/**
+ * Precise sharding algorithm to shard table.
+ *
+ * Sharding Sphere will use it to handle SQL with equal or in query condition.
+ */
+public final class PreciseTableShardingAlgorithm implements PreciseShardingAlgorithm<String> {
+    
+    private final static int TB_NUMBER = 10;
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseTableShardingAlgorithm.java]: Same question here. Can't inject the value?

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java (hunk 2 above)
Comment (wu-sheng) [reply oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java]: My guess, the actual meaning is, `Database Location Algorithm, locate the specific database instance in the given cluster`

Same questions for all `*Algorithm`.

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/RangeDatabaseShardingAlgorithm.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+import java.util.Collection;
+import java.util.Date;
+import java.util.LinkedHashSet;
+
+/**
+ * Range sharding algorithm to shard database.
+ *
+ * Sharding Sphere will use it to handle SQL with between query condition.
+ */
+public final class RangeDatabaseShardingAlgorithm implements RangeShardingAlgorithm<String> {
```
Comment (wu-sheng) [inline oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/RangeDatabaseShardingAlgorithm.java]: What is this for? And what is the difference between this and `PreciseDatabaseShardingAlgorithm`?

**Comment by wu-sheng** [comment]
> Hi, we need to setup e2e for shardingsphere before the main features to merge, right?

That is my concern. Traditionally, we don't merge PR without tests(including e2e), unless they are from experienced committer/PMC.

**Comment by wu-sheng** [comment]
@dmsolr You are doing storage UT framework and potential separated validator, right? I assume, without specific reasons, @tristaZero should finish the e2e by herself. This is an important step to understand how the SkyWalking community works, and help her to be one of qualified committer in the future.

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java (hunk 1 above)
Comment (tristaZero) [reply oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java]: Sorry that user can not inject the value into any implementation of this interface.

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java (hunk 2 above)
Comment (tristaZero) [reply oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/PreciseDatabaseShardingAlgorithm.java]: Precise sharding algorithm is a special item of Sharding Sphere. You can learn it more [Here](https://shardingsphere.apache.org/document/current/en/features/sharding/concept/sharding/). If necessary, i can add this description link to comment.

File: oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/RangeDatabaseShardingAlgorithm.java (hunk 4 above)
Comment (tristaZero) [reply oap-server/server-storage-plugin/storage-shardingsphere-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/shardingsphere/RangeDatabaseShardingAlgorithm.java]: Please note the small difference between `handle SQL with between(>= / <=) query condition` and `handle SQL with equal or in (= / IN) query condition`.
Sorry my comments confused you, do you have a idea to describe them better?

**Comment by tristaZero** [comment]
@wu-sheng @dmsolr 
> we don't merge PR without tests(including e2e)

Hi, thanks for your comment. I can provide unit tests for all the Algorithms, if what you said is a rule of SW community.

In addition, if others are interested and fimiliar with e2e than me, i am glad he or she can join me to finish this task togather.

**Comment by wu-sheng** [comment]
@tristaZero e2e is not that complex, as you have wrote one for MySQL. For shardingsphere, it is same as the prev one. What helps do you need? Your words make me feeling you have blocks?

**Comment by dmsolr** [comment]
Hi @tristaZero 
I am sorry. In fact, I think I can take little time to finish the e2e part if you provided the configuration of ShardSphere. In other words, we need to spend time to make ShardSphere work in a Docker container. 
If you don't want to waste volume time learning the e2e framework temporarily, I am willing to do it. :)

**Comment by dmsolr** [comment]
Oh, if you have written e2e-mysql, I think you are so familiar with e2e. @tristaZero

**Comment by tristaZero** [comment]
@wu-sheng @dmsolr Make sense to me.

**Comment by tristaZero** [comment]
Hi,

Since this PR has conflicting files and it will become a different solution with ShardingSphere incoming release, so I decided to close it. Please look forward to my consequent PR. 

Best,
Trista

**Pull request closed without merging**